programs/stake/src/stake_state.rs

```diff
@@ -414,29 +428,57 @@ impl Stake {
                 0
             };
 
-            total_rewards +=
-                (self.delegation.stake(*epoch, stake_history) * epoch_credits) as f64 * point_value;
+            points += u128::from(self.delegation.stake(*epoch, stake_history))
+                * u128::from(epoch_credits);
 
             // don't want to assume anything about order of the iterator...
             credits_observed = credits_observed.max(*credits);
         }
-        // don't bother trying to collect fractional lamports
-        if total_rewards < 1f64 {
+        (points, credits_observed)
+    }
+
+    /// for a given stake and vote_state, calculate what distributions and what updates should be made
+    /// returns a tuple in the case of a payout of:
+    ///   * staker_rewards to be distributed
+    ///   * voter_rewards to be distributed
+    ///   * new value for credits_observed in the stake
+    //  returns None if there's no payout or if any deserved payout is < 1 lamport
+    pub fn calculate_rewards(
+        &self,
+        point_value: &PointValue,
+        vote_state: &VoteState,
+        stake_history: Option<&StakeHistory>,
+    ) -> Option<(u64, u64, u64)> {
+        let (points, credits_observed) =
+            self.calculate_points_and_credits(vote_state, stake_history);
+
+        if points == 0 || point_value.points == 0 {
             return None;
         }
 
-        let (voter_rewards, staker_rewards, is_split) = vote_state.commission_split(total_rewards);
+        let rewards = points
+            .checked_mul(u128::from(point_value.rewards))
+            .unwrap()
+            .checked_div(point_value.points)
+            .unwrap();
+
+        assert!(rewards <= u128::from(std::u64::MAX));
+        let rewards = rewards as u64;
```

Comment:
nits: I think we can use `try_from` like this: `let rewards = u64::try_from(rewards).unwrap()`

Reply:
please feel free to push that update to my fork